code/modules/reagents/grinders.dm

```diff
@@ -0,0 +1,341 @@
+
+		if (!processing_chamber)
+			is_chamber_empty = 1
+			processing_chamber = "Nothing."
+		if (!beaker)
+			beaker_contents = "<B>No beaker attached.</B><br>"
+		else
+			is_beaker_ready = 1
+			beaker_contents = "<B>The beaker contains:</B><br>"
+			var/anything = 0
+			for(var/datum/reagent/R in beaker.reagents.reagent_list)
+				anything = 1
+				beaker_contents += "[R.volume] - [R.name]<br>"
+			if(!anything)
+				beaker_contents += "Nothing<br>"
+
+
+		dat += {"
+	<b>Processing chamber contains:</b><br>
+	[processing_chamber]<br>
+	[beaker_contents]<hr>
+	"}
+		if (is_beaker_ready && !is_chamber_empty && !(stat & (NOPOWER|BROKEN)))
+
+			// AUTOFIXED BY fix_string_idiocy.py
+			// C:\Users\Rob\\documents\\\projects\vgstation13\code\\modules\reagents\Chemistry-Machinery.dm:1016: dat += "<A href='?src=\ref[src];action=grind'>Grind the reagents</a><BR>"
+			dat += {"<A href='?src=\ref[src];action=process'>Process contents</a><BR>
+				<A href='?src=\ref[src];action=extract'>Extract reagents</a><BR><BR>"}
+			// END AUTOFIX
+		if(holdingitems && holdingitems.len > 0)
+			dat += "<A href='?src=\ref[src];action=eject'>Eject contents</a><BR>"
+		if (beaker)
+			dat += "<A href='?src=\ref[src];action=detach'>Detach beaker</a><BR>"
+	else
+		dat += "Please wait..."
+	dat = list2text(dat)
+	var/datum/browser/popup = new(user, "reagentgrinder", "All-In-One Grinder", src)
+	popup.set_content(dat)
+	popup.open()
+	onclose(user, "reagentgrinder")
+
+/obj/machinery/reagentgrinder/Topic(href, href_list)
+	if(..()) return 1
+	usr.set_machine(src)
+	switch(href_list["action"])
+		if ("process")
+			grind(1)
+		if("extract")
+			grind(0)
+		if("eject")
+			eject()
+		if ("detach")
+			detach()
+	src.updateUsrDialog()
+
+/obj/machinery/reagentgrinder/proc/detach()
+	if (usr.stat != 0) return
+	if (!beaker) return
+	beaker.forceMove(get_turf(src))
+	if(istype(beaker, /obj/item/weapon/reagent_containers/glass/beaker/large/cyborg))
+		var/mob/living/silicon/robot/R = beaker:holder:loc
+		if(R.module_state_1 == beaker || R.module_state_2 == beaker || R.module_state_3 == beaker)
+			beaker.forceMove(R)
+		else
+			beaker.forceMove(beaker:holder)
+	beaker = null
+	update_icon()
+
+/obj/machinery/reagentgrinder/proc/eject()
+	if (usr.stat != 0)
+		return
+	if (holdingitems && holdingitems.len == 0)
+		return
+
+	for(var/obj/item/O in holdingitems)
+		O.forceMove(get_turf(src))
+	holdingitems.Cut()
+
+/obj/machinery/reagentgrinder/proc/is_allowed(var/obj/item/I)
+	if(I.grindable_reagent || is_type_in_list(I, list(/obj/item/weapon/reagent_containers/food/snacks,/obj/item/weapon/grown))) return 1
+	return 0
+
+/obj/machinery/reagentgrinder/proc/grind(var/process = 0) //passed to grind_item
+	if (!beaker || (beaker && beaker.reagents.total_volume >= beaker.reagents.maximum_volume))
+		return
+	playsound(get_turf(src), speed_multiplier < 2 ? 'sound/machines/blender.ogg' : 'sound/machines/blenderfast.ogg', 50, 1)
+	inuse = 1
+	spawn(60/speed_multiplier)
```

Comment:
can't be do_after because that would require the user to stand by the grinder

that's not how a blender works